runtime/src/append_vec.rs

```diff
@@ -163,20 +163,29 @@ pub struct AccountMeta {
     /// lamports in the account
     pub lamports: u64,
     /// the epoch at which this account will next owe rent
-    pub rent_epoch: Epoch,
+    pub rent_epoch_or_application_fees: u64,
     /// the program that owns this account. If executable, the program that loads this account.
     pub owner: Pubkey,
     /// this account's data contains a loaded program (and is now read-only)
     pub executable: bool,
+    /// the epoch at which this account will next owe rent or application fees for the account
+    /// switched by the boolean above
+    pub has_application_fees: bool,
```

Comment:
We could potentially use something like bitflags to replace the current `bool` held by `executable`, in order to capture both: https://gist.github.com/rust-play/0616cd8f97bc8fd94718247c446250dd

This still breaks current sanitization checks in appendvec so it'd require the entire cluster to be on a version w/ the change. Has the advantage that it's backwards compatible with the current format though.

Reply:
@jeffwashington @brooksprumo or @yhchiang-sol: one or all of you might have more thoughts on how an update like this can/should be handled, or if it conflicts with your proposed new account storage.

Reply:
Yes i was thinking of the same thing, that I have to do some bitwise operations to do this. I just wanted a review from you guys about this. Thanks a lot, I will do the required changes.

Reply:
> Oops, I posted my review earlier today on the [other PR](https://github.com/blockworks-foundation/solana/pull/23#discussion_r1097795315)... Here's the text copied:
> 
> I think we can probably get the desired behavior without adding new fields. And really, we want/need to avoid new fields that end up the in actual append vec that's on disk. With a new type for handling rent-epoch-or-application-fee, plus some customer serialization/deserialization logic, we likely can implement this.
> 
> Note, this comment applies to the sdk `Account` as well; there will be a few differences, but I'll try and call those out as I see them.
> 
> The account on disk can't change yet, so we can't add more bytes (i.e. the flag to say rent-epoch-or-application-fee). We also know that rent is going away, but will not be gone for just under 2 years, which means we need to support rent epoch practically indefinitely. Right now we're at Rent Epoch 408. An upper bound for Rent Epoch in two years is ~773 (assuming one epoch per two days). `773` is a pretty small number to store in a u64, and we see there are plenty of spare bits. We could store rent epoch as a u16 and still have plenty of space left (like ~350 more years of epochs).
> 
> We also know most accounts are rent-exempt, and accounts are not allowed become rent-paying (minus one remaining known issue w.r.t. validators and rewards). We will be skipping rent collection on accounts known to be rent exempt. These accounts will have a sentinel value into their `Account`:
> 
> https://github.com/solana-labs/solana/blob/f3457a8485eca1bb034eb206f4ab947ffbc8906f/runtime/src/rent_collector.rs#L32-L35
> 
> And we could also say that Application Fees are only allowed on rent-exempt accounts. Thus, the rent epoch field can be reused. We could steal the top bit (or byte) to switch on Rent Epoch or Application Fee.
> 
> ```rust
> enum RentEpochOrApplicationFee {
>     RentEpoch(Epoch), // practically Epoch is stored as a u16
>     ApplicationFee(u56), // <-- max size could be a u63 if we want, too
> }
> ```
> 
> When rent collection asks for the rent epoch to see if this account is rent-exempt or not, if the enum variant is `ApplicationFee`, then it would return `RENT_EXEMPT_RENT_EPOCH`.
> 
> This would build on/require the work that sets `RENT_EXEMPT_RENT_EPOCH` to be activated, but I think that's reasonable.
> 
> Some additional things to work out:
> 
>     * Would need to manually implement the serialization/deserialization of `RentEpochOrApplicationFee` so that it fit within a u64
>       
>       * The in-memory representation could use [modular-bitfield](https://crates.io/crates/modular-bitfield) to help
> 
>     * Would need to ensure compatibility with older Snapshots that were created with the old Accounts
> 
>     * Maybe change the in-memory representation of Epoch to u16 first? This could make point_up easier
> 
> 
> Hope this helps! I know this impl is a lot of work, touching a lots of different parts of the validator code. This work and the SIMD will be helpful to the network!

Hey thanks a for the review. The idea is very good.

 I agree with you in case of append_vec that we should not add a new field to the AccountMeta. But I am confused for the `Account` struct. We have enough space between the field `executable` and `rent_epoch_and_application_fees` because of `#[repr(C)]` the executable is at offset 64 bytes and previous `rent_epoch` is at offset 72 bytes, so there are 8 bytes of data between them. I also added a unit test to test that serialization of the old account structure is compatible with deserialization of new account structure.
 
 I am curious to know why this wont work.
 
 I am more inclined towards Andrews idea of implementing bitwise operator for executable and has_application_fees just in append_vec so that we do not change the size of the stucture in the disk. And it is reasonable assumption that accounts with rent_epoch cannot have application fees.

Reply:
@apfitzge Updated as per your suggestion, added unit tests to test everything works as expected.